src/transformers/modeling_flash_attention_utils.py

```diff
@@ -203,6 +258,33 @@ def _flash_attention_forward(
             **flash_kwargs,
         )
         attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
+
+    # if position_ids is provided and check not all examples (row) contain only 1 sequence,
+    # then use `flash_attn_varlen_func` to prevent cross-example attention and also allow padding free approach
+    elif position_ids is not None and not (position_ids[:,-1]==position_ids.size(1)-1).all():
```

Comment:
this is an input dependent control flow, but we already don't support compile with FA2 AFAIK. 
Alright for now!